### PR TITLE
Existential Quantification

### PR DESCRIPTION
**Pull request by CodaFi**
An exploration (and hopefully, eventual solution) to #43.

### CODE REVIEW

**Comment by CodaFi** [comment]
I... Think I've done it!

**Comment by CodaFi** [comment]
OK, caveats:

This is not *real existential quantification*.  That would be impossible (or I'd at least win some kind of award).  The question now is do we treat a failing naïve search as a real failure or do the ScalaCheck thing and add an `Ambiguity` clause to `Result`?

File: SwiftCheck/Property.swift [hunk 1]
```diff
@@ -254,12 +276,25 @@ public enum Callback {
...
+/// The types of quantification SwiftCheck can perform.
+public enum Quantification {
+	/// Universal Quantification ("for all").
+	case Universal
+	/// Existential Quanfication ("there exists").
+	case Existential
+	/// Uniqueness Quantification ("there exists one and only one")
+	// case Uniqueness
+	/// Counting Quantification ("there exist exactly k")
+	// case Counting
```
Comment (robrix) [inline SwiftCheck/Property.swift]: If you do add these, I suggest you just do `Counting`. I don’t think `Uniqueness` adds much to the interface and, presumably the implementation of `Counting` would subsume `Uniqueness`.

File: SwiftCheck/Property.swift [hunk 2]
```diff
@@ -290,13 +326,15 @@ public struct TestResult {
 	let callbacks		: [Callback]
 	/// Indicates that any further testing of the property should cease.
 	let abort			: Bool
+	///
+	let quantifier		: Quantification
 
 	/// Destructures a test case into a matcher that can be used in switch statement.
-	public func match() -> TestResultMatcher {
-		return .MatchResult(ok: ok, expect: expect, reason: reason, theException: theException, labels: labels, stamp: stamp, callbacks: callbacks, abort: abort)
+	internal func match() -> TestResultMatcher {
```
Comment (robrix) [inline SwiftCheck/Property.swift]: These are all `internal` by default, no? (To be clear, I am not intending to be prescriptive re: code style, just curious why it should be explicit on the function & `init` but not the `let`s above.)

File: SwiftCheck/Test.swift [hunk 3]
```diff
@@ -102,6 +102,40 @@ public func forAll<A : Arbitrary, B : Arbitrary, C : Arbitrary, D : Arbitrary, E
...
+/// The nature of Existential Quantification means we have to enumerate over the entire domain of 
+/// `A` in order to return a proper value.  Because such a traversal is both impractical and leads 
+/// to computationally questionable behavior (infinite loops and the like), SwiftCheck instead
+/// interprets `exists` as a finite search over arbitrarily many values (around 500).  No shrinking
+/// is performed during the search.
+///
+/// It is recommended that you avoid existential quantification and instead reduce your property to
+/// `Skolem Normal Form <https://en.wikipedia.org/wiki/Skolem_normal_form/>`_.  `SNF` involves 
+/// turning every `exists` into a function returning the existential value, taking any other 
+/// parameters being quantified over as needed.
```
Comment (robrix) [inline SwiftCheck/Test.swift]: When might `exists` be more appropriate than SNF?

**Comment by CodaFi** [comment]
Welp, this is kinda strange... Travis doesn't seem to think this branch exists.  I'll try again in a different pull request.

**Pull request closed without merging**